sys/include/usb/usbus.h

```diff
@@ -0,0 +1,515 @@
+    usbdev_ep_t *ep;             /**< ptr to the matching usbdev endpoint */
+    uint16_t maxpacketsize;      /**< Max packet size of this endpoint */
+    uint8_t interval;            /**< Poll interval for interrupt endpoints */
+    bool active;                 /**< If the endpoint should be activated after
+                                      reset */
+} usbus_endpoint_t;
+
+/**
+ * @brief USBUS interface alternative setting
+ *
+ * Used for specifying alternative interfaces for an @ref usbus_interface_t
+ */
+typedef struct usbus_interface_alt {
+    struct usbus_interface_alt *next; /**< Next alternative setting */
+    usbus_hdr_gen_t *hdr_gen;         /**< Optional additional header
+                                           generator */
+    usbus_endpoint_t *ep;             /**< List of associated endpoints for
+                                           this alternative setting */
+} usbus_interface_alt_t;
+
+/**
+ * @brief USBUS interface
+ */
+typedef struct usbus_interface {
+    struct usbus_interface *next;     /**< Next interface (set by USBUS during
+                                           registration)                     */
+    struct usbus_interface_alt *alts; /**< List of alt settings              */
+    usbus_handler_t* handler;         /**< Handlers for this interface       */
+    usbus_hdr_gen_t *hdr_gen;         /**< Optional additional header
+                                           generators                        */
+    usbus_endpoint_t *ep;             /**< Linked list of endpoints belonging
+                                           to this interface                 */
+    usbus_string_t *descr;            /**< Descriptor string                 */
+    uint16_t idx;                     /**< Interface index, (set by USBUS
+                                           during registration               */
+    uint8_t class;                    /**< USB interface class               */
+    uint8_t subclass;                 /**< USB interface subclass            */
+    uint8_t protocol;                 /**< USB interface protocol            */
+} usbus_interface_t;
+
+/**
+ * @brief USBUS event handler function pointers
+ */
+typedef struct usbus_handler_driver{
+
+    /**
+     * @brief Initialize the event handler
+     *
+     * This function is called in the USBUS thread context to initialize the
+     * event handler
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     */
+    void (*init)(usbus_t *usbus, struct usbus_handler *handler);
+
+    /**
+     * @brief event handler function
+     *
+     * This function is passed USBUS events
+     *
+     * @param usbus     USBUS context
+     * @param handler   handler context
+     * @param event     @ref USBUS_MSG_TYPES "event" to handle
+     * @param arg       Additional argument
+     */
+    int (*event_handler)(usbus_t *usbus, struct usbus_handler *handler, uint16_t event, void *arg);
```

Comment:
One flaw of the API I've found so far is that too many event-like operations are multiplexed over this single callback. Splitting this into multiple callbacks {events, transfer, setup} reduces the API complexity, allows for not handling certain events (leaving the function ptr at `NULL`) and prevents every handler from having [switch-case statements](https://github.com/RIOT-OS/RIOT/pull/11085/files#diff-4f2c0a81a745c2641ad7a3c388997c8dR313) to demultiplex the reported event

Reply:
Splitted these, documentation can be improved and a few defines should be changed, but so far I'm happy with it and it reduces code size a bit